ipfs-retriever/lib/store.js

```diff
@@ -0,0 +1,392 @@
+ *   retrieval
+ * @returns {Promise<void>} - A promise that resolves when the log is inserted.
+ */
+export async function logRetrievalResult(env, params) {
+  console.log('retrieval log', params)
+  const {
+    cacheMiss,
+    egressBytes,
+    responseStatus,
+    timestamp,
+    performanceStats,
+    requestCountryCode,
+    dataSetId,
+  } = params
+
+  try {
+    await env.DB.prepare(
+      `
+      INSERT INTO retrieval_logs (
+        timestamp,
+        response_status,
+        egress_bytes,
+        cache_miss,
+        fetch_ttfb,
+        fetch_ttlb,
+        worker_ttfb,
+        request_country_code,
+        data_set_id
+      )
+      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
+      `,
+    )
+      .bind(
+        timestamp,
+        responseStatus,
+        egressBytes,
+        cacheMiss,
+        performanceStats?.fetchTtfb ?? null,
+        performanceStats?.fetchTtlb ?? null,
+        performanceStats?.workerTtfb ?? null,
+        requestCountryCode,
+        dataSetId,
+      )
+      .run()
+  } catch (error) {
+    console.error(`Error inserting log: ${error}`)
+    // TODO: Handle specific SQL error codes if needed
+    throw error
+  }
+}
+
+/**
+ * Validates query results and returns provider info. This is a shared helper
+ * used by both getStorageProviderAndValidatePayerByWalletAndCid and
+ * getStorageProviderAndValidatePayerByDataSetAndPiece.
+ *
+ * @param {object} params
+ * @param {any[]} params.results - The query results to validate
+ * @param {string} params.payerAddress - The address of the client paying for
+ *   the request
+ * @param {string} params.lookupKey - Descriptive key for error messages (e.g.,
+ *   "IPFS Root CID 'bafk...'")
+ * @returns {{
+ *   serviceProviderId: string
+ *   serviceUrl: string
+ *   dataSetId: string
+ *   pieceId: string
+ *   ipfsRootCid: string
+ * }}
+ */
+function validateQueryResultsAndGetProvider(params) {
+  const { results, payerAddress, lookupKey } = params
+
+  httpAssert(
+    results && results.length > 0,
+    404,
+    `${lookupKey} does not exist or may not have been indexed yet.`,
+  )
+
+  const withServiceProvider = results.filter(
+    (row) => row && row.service_provider_id != null,
+  )
+  httpAssert(
+    withServiceProvider.length > 0,
+    404,
+    `${lookupKey} exists but has no associated service provider.`,
+  )
+
+  const withPaymentRail = withServiceProvider.filter(
+    (row) =>
+      row.payer_address && row.payer_address.toLowerCase() === payerAddress,
+  )
+  httpAssert(
+    withPaymentRail.length > 0,
+    402,
+    `There is no Filecoin Warm Storage Service deal for payer '${payerAddress}' and ${lookupKey}.`,
+  )
+
+  const withCDN = withPaymentRail.filter(
+    (row) => row.with_cdn && row.with_cdn === 1,
+  )
+  httpAssert(
+    withCDN.length > 0,
+    402,
+    `The Filecoin Warm Storage Service deal for payer '${payerAddress}' and ${lookupKey} has withCDN=false.`,
+  )
```

Comment:
The SQL query and some of the following filters are specific to IPFS retrievals, but many filters are shared with the Piece retrievals. It makes me wonder if it's worth extracting this into a shared helper?